### PR TITLE
Core|ORC|Spark: Remove deprecated functionality

### LINKED ISSUE #6155: Remove API deprecations for 1.2.0
### Feature Request / Improvement

iceberg-api / iceberg-core have a few deprecated methods that we should remove before releasing 1.2.0

### Query engine

_No response_

### PR DESCRIPTION
**Pull request by nastra**
fixes #6155

### CODE REVIEW

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteManifestsSparkAction.java [hunk 1]
```diff
@@ -356,7 +356,7 @@ private static ManifestFile writeManifest(
         long snapshotId = row.getLong(0);
         long sequenceNumber = row.getLong(1);
         Row file = row.getStruct(2);
-        writer.existing(wrapper.wrap(file), snapshotId, sequenceNumber);
+        writer.existing(wrapper.wrap(file), snapshotId, sequenceNumber, null);
```
Comment (nastra) [inline spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteManifestsSparkAction.java:359]: @aokolnychyi would we need to backport https://github.com/apache/iceberg/pull/6193 to Spark 2.4 here?

File: core/src/main/java/org/apache/iceberg/ManifestEntry.java [hunk 2]
```diff
@@ -103,27 +103,6 @@ default boolean isLive() {
...
-   * Returns the data sequence number of the file if the entry status is ADDED or EXISTING.
-   * Otherwise, returns the sequence number of the snapshot in which the file was removed.
-   *
-   * <p>Note that usage of this method should be avoided as it behaves inconsistently for different
-   * entry statutes. Use {@link #dataSequenceNumber()} instead.
-   *
-   * @deprecated since 1.0.0, will be removed in 1.1.0; use {@link #dataSequenceNumber()} instead.
-   */
-  @Deprecated
-  Long sequenceNumber();
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/ManifestEntry.java:116]: @aokolnychyi could you please review all places that adjust sequenceNumber -> dataSequenceNumber?

File: core/src/main/java/org/apache/iceberg/GenericManifestEntry.java [hunk 3]
```diff
@@ -30,7 +30,6 @@
   private Status status = Status.EXISTING;
   private Long snapshotId = null;
   private Long dataSequenceNumber = null;
-  private Long sequenceNumber = null;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/GenericManifestEntry.java:33]: @aokolnychyi, can you take a look at these changes?

File: core/src/main/java/org/apache/iceberg/LocationProviders.java [hunk 4]
```diff
@@ -84,10 +84,7 @@ static class DefaultLocationProvider implements LocationProvider {
     private static String dataLocation(Map<String, String> properties, String tableLocation) {
       String dataLocation = properties.get(TableProperties.WRITE_DATA_LOCATION);
       if (dataLocation == null) {
-        dataLocation = properties.get(TableProperties.WRITE_FOLDER_STORAGE_LOCATION);
-        if (dataLocation == null) {
-          dataLocation = String.format("%s/data", tableLocation);
-        }
+        dataLocation = String.format("%s/data", tableLocation);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:87]: Why are we removing public-facing functionality like this? I don't think that we should ever drop support for this because it will break existing tables.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 5]
```diff
@@ -758,20 +758,6 @@ protected Map<String, String> summary() {
...
-   * Apply the update's changes to the base table metadata and return the new manifest list.
-   *
-   * @param base the base table metadata to apply changes to
-   * @return a manifest list for the new snapshot.
-   * @deprecated Will be removed in 1.2.0, use {@link MergingSnapshotProducer#apply(TableMetadata,
-   *     Snapshot)}.
-   */
-  @Deprecated
-  @Override
-  public List<ManifestFile> apply(TableMetadata base) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:771]: Why remove these? It doesn't seem like a big deal to keep the versions that default the parent snapshot to current.

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 6]
```diff
@@ -231,15 +231,8 @@ private TableProperties() {}
   public static final String OBJECT_STORE_ENABLED = "write.object-storage.enabled";
   public static final boolean OBJECT_STORE_ENABLED_DEFAULT = false;
 
-  /** @deprecated Use {@link #WRITE_DATA_LOCATION} instead. */
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:234]: Ah, I think this was misleading. This is telling people not to use the property, but it doesn't mean that we can remove the property. As I said in a comment above, this will break older tables so there's no good reason to ever remove it.

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 7]
```diff
@@ -342,19 +335,6 @@ private TableProperties() {}
...
-   *     0.13.0.
-   */
-  @Deprecated
-  public static final String MERGE_CARDINALITY_CHECK_ENABLED =
-      "write.merge.cardinality-check.enabled";
-  /**
-   * @deprecated will be removed in 0.14.0, the cardinality check is always performed starting from
-   *     0.13.0.
-   */
-  @Deprecated public static final boolean MERGE_CARDINALITY_CHECK_ENABLED_DEFAULT = true;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:356]: This is not the same situation as the storage location properties. This should be removed because it no longer does anything.

File: core/src/main/java/org/apache/iceberg/deletes/Deletes.java [hunk 8]
```diff
@@ -83,21 +83,6 @@ public static <T> CloseableIterable<T> markDeleted(
...
-   * Returns the remaining rows (the ones that are not deleted), while counting the deleted ones.
-   *
-   * @param rows the rows to process
-   * @param isDeleted a predicate that determines if a row is deleted
-   * @return the processed rows
-   * @deprecated Will be removed in 1.2.0, use {@link Deletes#filterDeleted(CloseableIterable,
-   *     Predicate, DeleteCounter)}.
-   */
-  @Deprecated
-  public static <T> CloseableIterable<T> filterDeleted(
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/deletes/Deletes.java:96]: I guess this is okay? It doesn't seem bad to have a version that doesn't require a counter.

File: data/src/test/java/org/apache/iceberg/data/orc/TestOrcRowIterator.java [hunk 9]
```diff
@@ -74,7 +75,7 @@ public void writeFile() throws IOException {
             .schema(DATA_SCHEMA)
             // write in such a way that the file contains 2 stripes each with 4 row groups of 1000
             // rows
-            .set("iceberg.orc.vectorbatch.size", "1000")
+            .set(TableProperties.ORC_WRITE_BATCH_SIZE, "1000")
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/data/orc/TestOrcRowIterator.java:78]: Is this change needed? It looks unrelated to the deprecation changes.

Comment (rdblue) [reply data/src/test/java/org/apache/iceberg/data/orc/TestOrcRowIterator.java:78]: Or is this a different property? I see the matching constant was removed.

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java [hunk 10]
```diff
@@ -193,11 +183,6 @@ public <D> FileAppender<D> build() {
         this.conf.set(entry.getKey(), entry.getValue());
       }
 
-      // for compatibility
-      if (conf.get(VECTOR_ROW_BATCH_SIZE) != null && config.get(ORC_WRITE_BATCH_SIZE) == null) {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/ORC.java:197]: What is the value in removing this? I think we should probably just keep it.

File: spark/v3.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala [hunk 11]
```diff
@@ -228,13 +225,6 @@ case class RewriteMergeInto(spark: SparkSession) extends Rule[LogicalPlan] with
     }
   }
 
-  private def isCardinalityCheckEnabled(table: Table): Boolean = {
-    PropertyUtil.propertyAsBoolean(
-      table.properties(),
-      MERGE_CARDINALITY_CHECK_ENABLED,
```
Comment (rdblue) [inline spark/v3.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala:234]: Okay, it looks like this property is still used by Spark 3.1, so we shouldn't actually remove it (or the ability to disable it!) until we remove support for 3.1.

File: spark/v3.1/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java [hunk 12]
```diff
@@ -370,48 +369,6 @@ public void testMergeWithMultipleUpdatesForTargetRow() {
         sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", tableName));
   }
 
-  @Test
-  public void testMergeWithDisabledCardinalityCheck() {
```
Comment (rdblue) [inline spark/v3.1/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java:374]: I think this is still valid for 3.1

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkParquetReader.java [hunk 13]
```diff
@@ -144,16 +143,16 @@ public void testInt96TimestampProducedBySparkIsReadCorrectly() throws IOExceptio
...
-        new ParquetWriteAdapter<>(
-            new NativeSparkWriterBuilder(outputFile)
-                .set("org.apache.spark.sql.parquet.row.attributes", sparkSchema.json())
-                .set("spark.sql.parquet.writeLegacyFormat", "false")
-                .set("spark.sql.parquet.outputTimestampType", "INT96")
-                .set("spark.sql.parquet.fieldId.write.enabled", "true")
-                .build(),
-            MetricsConfig.getDefault())) {
-      writer.addAll(rows);
+    try (ParquetWriter<InternalRow> writer =
```
Comment (rdblue) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkParquetReader.java:146]: Why were these changes needed?

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteManifestsSparkAction.java (hunk 1 above)
Comment (nastra) [reply spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteManifestsSparkAction.java:359]: I've opened https://github.com/apache/iceberg/pull/6289 to do that

File: data/src/test/java/org/apache/iceberg/data/orc/TestOrcRowIterator.java (hunk 9 above)
Comment (nastra) [reply data/src/test/java/org/apache/iceberg/data/orc/TestOrcRowIterator.java:78]: `iceberg.orc.vectorbatch.size` was deprecated for quite some time. I've restored all code that removes deprecated properties and code that deals with legacy properties

File: spark/v3.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala (hunk 11 above)
Comment (nastra) [reply spark/v3.1/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala:234]: I've restored this and updated the comment on the property itself to mention that we will drop it once Spark 3.1 support will be dropped

File: core/src/main/java/org/apache/iceberg/deletes/Deletes.java (hunk 8 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/deletes/Deletes.java:96]: This is functionality that was adjusted by #4588. We then re-introduced + deprecated the method in #6146 to not have breaking API changes. Also there's currently no place in the code that would use this, so it seems better to remove it?

File: core/src/main/java/org/apache/iceberg/TableProperties.java (hunk 7 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/TableProperties.java:356]: I've restored this property and mentioned that it can only be dropped once Spark 3.1 support is dropped

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 5 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:771]: This is code that changed this method in #4926. We then restored+deprecated the old API in #6146 for one minor release and are removing it here now. 
Let me know if we'd want to keep those methods, then we can just drop the `Deprecated` annotations from those

File: core/src/main/java/org/apache/iceberg/LocationProviders.java (hunk 4 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/LocationProviders.java:87]: I've restored this

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkParquetReader.java (hunk 13 above)
Comment (nastra) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkParquetReader.java:146]: this was using the deprecated `ParquetWriteAdapter` class, so I started changing code that uses this class so that we can eventually remove `ParquetWriteAdapter`

File: spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteManifestsSparkAction.java (hunk 1 above)
Comment (aokolnychyi) [reply spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteManifestsSparkAction.java:359]: Yep, thanks. I merged.

**Comment by aokolnychyi** [comment]
I'll take a look at seq number related changes today.

File: core/src/main/java/org/apache/iceberg/GenericManifestEntry.java [hunk 14]
```diff
@@ -154,12 +140,6 @@ public void setSnapshotId(long newSnapshotId) {
   public void setDataSequenceNumber(long newDataSequenceNumber) {
     // always reset sequenceNumber whenever dataSequenceNumber is changed
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/GenericManifestEntry.java:141]: Shall we remove this comment as well? It no longer applies as we removed `sequenceNumber`.

File: core/src/main/java/org/apache/iceberg/GenericManifestEntry.java [hunk 15]
```diff
@@ -180,7 +160,6 @@ public void put(int i, Object v) {
       case 2:
         // always reset sequenceNumber whenever dataSequenceNumber is changed
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/GenericManifestEntry.java:161]: Here as well.

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 16]
```diff
@@ -343,15 +343,15 @@ private TableProperties() {}
   public static final String MERGE_MODE_DEFAULT = RowLevelOperationMode.COPY_ON_WRITE.modeName();
 
   /**
-   * @deprecated will be removed in 0.14.0, the cardinality check is always performed starting from
-   *     0.13.0.
+   * @deprecated will be removed once Spark 3.1 support is dropped, the cardinality check is always
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:346]: Makes sense to me.

**Review by aokolnychyi — APPROVED** [review]
LGTM, I'd just remove two comments that no longer apply. Thanks, @nastra!

File: spark/v3.1/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java (hunk 12 above)
Comment (nastra) [reply spark/v3.1/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java]: I've restored this

File: core/src/main/java/org/apache/iceberg/GenericManifestEntry.java (hunk 14 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/GenericManifestEntry.java:141]: good call, removed it

File: core/src/main/java/org/apache/iceberg/GenericManifestEntry.java (hunk 15 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/GenericManifestEntry.java:161]: good call, removed it

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @nastra!
